src/sources/readaheadframebuffer.cpp

```diff
@@ -0,0 +1,477 @@
+                << bufferedRange()
+                << "and input buffer"
+                << inputRange;
+        switch (discontinuityGapMode) {
+        case DiscontinuityGapMode::Skip:
+            reset(inputRange.start());
+            break;
+        case DiscontinuityGapMode::FillWithSilence: {
+            const auto clearFrameCount = gapRange.length();
+            adjustCapacityBeforeBuffering(clearFrameCount);
+            const auto clearSampleCount =
+                    m_signalInfo.frames2samples(clearFrameCount);
+            const SampleBuffer::WritableSlice writableSamples(
+                    m_sampleBuffer.growForWriting(clearSampleCount));
+            DEBUG_ASSERT(writableSamples.length() == clearSampleCount);
+            SampleUtil::clear(
+                    writableSamples.data(),
+                    clearSampleCount);
+        } break;
+        default:
+            DEBUG_ASSERT(!"Unknown DiscontinuityGapMode");
+        }
+    }
+
+    DEBUG_ASSERT(writeIndex() == inputRange.start());
+    if (inputRange.empty()) {
+        return inputBuffer;
+    }
+    // Consume the readable sample data by copying it into the internal buffer
+#if VERBOSE_DEBUG_LOG
+    kLogger.debug()
+            << "Buffering sample frames"
+            << inputRange;
+#endif
+    adjustCapacityBeforeBuffering(inputRange.length());
+    const auto copySampleCount =
+            m_signalInfo.frames2samples(inputRange.length());
+    const SampleBuffer::WritableSlice writableSamples(
+            m_sampleBuffer.growForWriting(copySampleCount));
+    DEBUG_ASSERT(writableSamples.length() == copySampleCount);
+    SampleUtil::copy(
+            writableSamples.data(),
+            inputBuffer.readableData(),
+            copySampleCount);
+    pInputSamples += copySampleCount;
+    inputRange.shrinkFront(inputRange.length());
+    DEBUG_ASSERT(inputRange.empty());
+    return ReadableSampleFrames(
+            inputRange,
+            SampleBuffer::ReadableSlice(
+                    pInputSamples,
+                    m_signalInfo.frames2samples(inputRange.length())));
+}
+
+WritableSampleFrames ReadAheadFrameBuffer::drainBuffer(
+        WritableSampleFrames outputBuffer) {
+    DEBUG_ASSERT(isValid());
+    auto outputRange = outputBuffer.frameIndexRange();
+#if VERBOSE_DEBUG_LOG
+    kLogger.debug()
+            << "drainBuffer:"
+            << "bufferedRange()"
+            << bufferedRange()
+            << "outputRange"
+            << outputRange;
+#endif
+    if (isEmpty() || outputBuffer.writableSlice().empty()) {
+        return outputBuffer;
+    }
+    DEBUG_ASSERT(isReady());
+    if (outputRange.start() < m_readIndex) {
+        // Buffered data starts beyond the requested range
+        return outputBuffer;
+    }
+    const auto consumableRange = intersect2(bufferedRange(), outputRange);
+    DEBUG_ASSERT(!consumableRange || *consumableRange <= outputRange);
+    if (!consumableRange || consumableRange->empty()) {
```

Comment:
This condition reveals that it is redundant to return an std::optional from intersect2.
The check for !consumableRange looks already like a validity check, but it is not it only checks if the Range itself is valid and the second check can be performed. This is redundant and IMHO dangerous.

If you like to have the convenience of a "!consumableRange", we can simply add a bool operator. 

The involved redirection has also performance penalty compared to returning the object by value, because it has also 16 byte, the same as the std::optional and in the valid case extra memory allocation is required. 

So I think we should revert this change and not use std::optional.

Reply:
The intersection of two completely disconnected intervals is undefined. In this case `std::nullopt` is the appropriate result. Returning an empty range with an arbitrary index (= 0) like the legacy function `intersect()` does is conceptually wrong and already caused a bug.

I refuse to revert back to the previous, incorrect implementation. Yes, in this special case the caller has to check the result for both cases. But making the intersection function itself incorrect to "optimize" it for this special use case would be wrong.

Reply:
The documentation of intersect2() describes in detail how the function works.

Reply:
Trading correct code for optimized but wrong code is a bad deal.

Reply:
This code is called often and it feels really wrong call malloc because you like to use std::optional. 
std::optional does make sense if the payload is less then 8 bytes or if the data has no way to express the invalid state or if default constructing is heavy. All this does not apply here. 

In such bussy code we should take care that all memory is allocated during initialization. 

In this case here it would be easy to define an internal state that distinguishes a zero length range with a valid offset from a invalid range. 

By the way intersect2() does not distinguish this yet. So probably we can use use a  zero length range and fix the bug you mention elsewhere.

Reply:
No memory is allocated here apart from some bytes on the stack during the function call. What do you refer to?

I disagree and will not change the code back to the state where the different results are indistinguishable. This would be error prone. We should have learned a lesson and not make the same mistake twice.

The overhead of std::optional is neglibile. We already made this decision once for RgbColor, i.e. use what the standard provides in favor of an optimized but custom solution.

Reply:
...and for RgbColor we even had some unused bits available to distinguish invalid from valid values.

Reply:
> No memory is allocated here apart from some bytes on the stack during the function call. What do you refer to?

It looks I have tricked myself, when checking the sizeof the std::optional compared with the playload. Today I did the check again and it turns out that the std::optional is always 4 byte greater than the payload. No heap memory allocation involved. 

But my general concern is still valid. Using the std::optional her complicates the code because all of the extra checks and have no benefit because we don't need to distinguish the null and empty case. 

Here is good answer that can be used as a rule of thumb: 
https://topanswers.xyz/cplusplus?q=923#a1085

Here we ave already an empty state that is sufficient, no need to complicate to code by extra null checks.

Reply:
RgbColor the case is different, because we have implement the assumption that the color is always valid. This is not the case here. 
Looking again at the RgbColor code, it would have been easy in the same way to just check the validity on the fly instead of lock down all places where the values can become invalid.
I consider our current approach as more error phone, because it needs assertions instead of being naturally valid.

Reply:
I can only repeat myself:
- We need to distinguish between *undefined* and *empty*, even if in some use cases those states are handled in the same way
- There is no single *empty* value, because every range has a start index that serves as the *anchor*. Returning `std::nullopt` enforces callers to explicitly handle *undefined* values instead of accidentally using the arbitrary *start* index of an *empty* range.

I refuse to write unsafe and unsound code.

Reply:
Using optional for function arguments and return types is actually the prime example. And here we are not even discussing about choosing a nullable pointer (with a single, well-defined nullptr value that cannot be misinterpreted) vs. optional.

Reply:
Sorry, Daniel, but I am not going to repeat my arguments once again. Then let me close this PR and instead focus on things where I can realize my vision and ideas of safe software engineering.